### PR TITLE
Introduce locking and unlocking in refund flow to prevent double refund due to race condition

### LINKED ISSUE #838: Double refund
#### Affected ticket(s)

N/A

#### What I expected

When I issue a refund, it should only be refunded once. 

#### What happened instead

The refund doubled up

#### Steps to reproduce the issue

I'm running an WP.com ecomm plan, admining the order through WP-admin - Stripe is in test mode. I'm using card 4242 4242 4242 4242

I placed an order on the front end for one product with multiple items, then visited the edit order screen:

Screen Shot 2019-04-11 at 1 12 08 PM

I refunded 1 item:

Screen Shot 2019-04-11 at 1 16 24 PM

I clicked "Refund 98 cents via stripe" which resulted in a double refund

Screen Shot 2019-04-11 at 1 19 07 PM

I also see a double refund in my order notes, although the refund ID appears to be the same:

Screen Shot 2019-04-11 at 1 21 23 PM

-------------------

- [ ] Issue assigned to next milestone.
- [ ] Issue assigned a priority (will be assessed by maintainers).

### PR DESCRIPTION
**Pull request by Mayisha**
Fixes #838 

## Changes proposed in this Pull Request:
- Introduced `lock_order_refund` and `unlock_order_refund` methods. These are similar to the `lock_order_payment` and `unlock_order_payment` methods.
- Adding lock to the order during the `process_refund` flow to prevent double refund due to webhook event coming before the `refund` API response.

## Testing instructions
- As a shopper place an order.
- Add a 10 seconds sleep ( `sleep(10)` ) [here](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/abstracts/abstract-wc-stripe-payment-gateway.php#L1167) after sending the refund request to create a fake race condition. This way the webhook will be processed before the `_stripe_refund_id` is set by the `process_refund` function.
- As an admin go to the order edit page and refund the order with a note.
- In `develop` branch notice that there are 2 refund notes.
- In this branch, confirm that there is only one note.

### CODE REVIEW

File: includes/abstracts/abstract-wc-stripe-payment-gateway.php [hunk 1]
```diff
@@ -1078,7 +1078,7 @@ public function update_fees( $order, $balance_transaction_id ) {
 	 * @return bool True or false based on success.
 	 * @throws Exception Throws exception when charge wasn't captured.
 	 */
-	public function process_refund( $order_id, $amount = null, $reason = '' ) {
+	public function process_refund( $order_id, $amount = null, $reason = '' ) { // mayisha
```
Comment (mattallan) [inline includes/abstracts/abstract-wc-stripe-payment-gateway.php]: Debugging comment I think

File: includes/abstracts/abstract-wc-stripe-payment-gateway.php [hunk 2]
```diff
@@ -1165,8 +1167,11 @@ public function process_refund( $order_id, $amount = null, $reason = '' ) {
 			);
 		}
 
+		sleep( 10 ); // Prevent
```
Comment (mattallan) [inline includes/abstracts/abstract-wc-stripe-payment-gateway.php]: Debugging code made it in the PR 😅

**Review by mattallan — APPROVED** [review]
Nice fix here @Mayisha :)

This PR is missing a changelog entry and I left a comment about some debugging code that made it in 😅 

- [x] Confirmed bug on `develop`: image
- [x] Tested refunds via Woo Admin
- [x] Tested refunds via Stripe dashboard

FWIW I was attempting to try and still get duplicate refunds by refunding on Stripe and then quickly at Woo at the same time. I wasn't able to get it to break so I think you'd have to be incredibly quick 

A couple of small things to fix, but nothing blocking. So happy to approve this.

File: includes/abstracts/abstract-wc-stripe-payment-gateway.php (hunk 1 above)
Comment (Mayisha) [reply includes/abstracts/abstract-wc-stripe-payment-gateway.php]: I keep forgetting my debug traces everytime 🤣

**Comment by buy** [comment]
Thank you all for fixing this!!!!!! 🎉 👏 💯

**Pull request merged**